lncore/addr.go

```diff
@@ -0,0 +1,319 @@
+	// Pkh is the pubkey hash as bech32.
+	Pkh    *string
+	
+	// Pubkey is the pubkey, as bytes, decoded from source format.
+	Pubkey []byte
+
+	// IPAddr is the IP address, if present.
+	IPAddr *string
+
+	// Port should always be defined, default is LnDefaultPort.
+	Port   uint16
+}
+
+// HasFullAddress returns whether or not we need to do a tracker/DHT lookup in
+// order to resolve the connection address for this address data.  Basically, it
+// checks if there's an IP address defined.
+func (d *LnAddressData) HasFullAddress() bool {
+	return d.IPAddr != nil
+}
+
+// AddrFmtFull is <bech32 pkh>:<hex pk>@<ip>:<port>
+const AddrFmtFull = "full"
+
+// AddrFmtFullBech32Pk is <bech32 pkh>:<bech32 pk>@<ip>:<port>
+const AddrFmtFullBech32Pk = "full_bech32_pk"
+
+// AddrFmtFullNoPort is <bech32 pkh>:<hex pk>@<ip>
+const AddrFmtFullNoPort = "full_no_port"
+
+// AddrFmtFullBech32PkNoPort is <bech32 pkh>:<bech32 pk>@<ip>
+const AddrFmtFullBech32PkNoPort = "full_bech32_pk_no_port"
+
+// AddrFmtBech32 is just <bech32 pkh>
+const AddrFmtBech32 = "bech32_pkh"
+
+// AddrFmtLit is just AddrFmtBech32
+const AddrFmtLit = "lit"
+
+// AddrFmtPubkey is <hex pubkey>
+const AddrFmtPubkey = "hex_pk"
+
+// AddrFmtBech32Pubkey is <bech32 pk>
+const AddrFmtBech32Pubkey = "bech32_pk"
+
+// AddrFmtPubkeyIP is <hex pk>@<ip>
+const AddrFmtPubkeyIP = "hex_pk_ip"
+
+// AddrFmtPubkeyIPPort is <hex pk>@<ip>:<port>
+const AddrFmtPubkeyIPPort = "hex_pk_ip_port"
+
+// AddrFmtLitFull is <bech32 pkh>@<ip>:<port>
+const AddrFmtLitFull = "lit_full"
+
+// AddrFmtLitIP is <bech32 pkh>@<ip>
+const AddrFmtLitIP = "lit_ip"
+
+// DumpAddressFormats returns the addresses in all of the formats that fully represent it.
+func DumpAddressFormats(data *LnAddressData) (map[string]string, error) {
+
+	ret := make(map[string]string)
+
+	if data.Port != LnDefaultPort && data.IPAddr == nil {
+		return ret, fmt.Errorf("nondefault port specified but IP not specified")
+	}
+
+ 	// Full.
+	if data.Pkh != nil && data.Pubkey != nil && data.IPAddr != nil {
+		ret[AddrFmtFull] = fmt.Sprintf("%s:%X@%s:%d", *data.Pkh, data.Pubkey, *data.IPAddr, data.Port)
+		ret[AddrFmtFullBech32Pk] = fmt.Sprintf("%s:%s@%s:%d", *data.Pkh, bech32.Encode(PubkeyBech32Prefix, data.Pubkey), *data.IPAddr, data.Port)
+	}
+	
+	// Minimal Lit.
+	if data.Pkh != nil {
+		ret[AddrFmtLit] = *data.Pkh
+	}
+
+	// Long Lit.
+	if data.Pkh != nil && data.IPAddr != nil {
+		if data.Port == LnDefaultPort {
+			ret[AddrFmtLitIP] = fmt.Sprintf("%s@%s", *data.Pkh, *data.IPAddr)
+		}
+		ret[AddrFmtLitFull] = fmt.Sprintf("%s@%s:%d", *data.Pkh, *data.IPAddr, data.Port)
+	}
+
+	// Hex pubkey stuff
+	if data.Pubkey != nil {
+		ret[AddrFmtPubkey] = fmt.Sprintf("%X", data.Pubkey)
+		if data.IPAddr != nil {
+			if data.Port == LnDefaultPort {
+				ret[AddrFmtPubkeyIP] = fmt.Sprintf("%X@%s", data.Pubkey, *data.IPAddr)
+			}
+			ret[AddrFmtPubkeyIPPort] = fmt.Sprintf("%X@%s:%d", data.Pubkey, *data.IPAddr, data.Port)
+		}
+	}
+
+	// TODO more
```

Comment:
I think these formats are enough actually, we just need two formats (maybe three, if you count just the pkh as one format).
1. Lit Addr (pkh) with Port
2. BOLT Addr (hex pubkey) with Port
(3). Lit Addr (pkh)

Again, dependent on how we would use this since displaying addresses requires only the two formats above unless I'mm missing something here.

Reply:
Yeah people would probably only end up using those 3 formats often, but since people do weird things I thought it makes sense to include all possible variations?  Especially since it's not hard for a person to manually convert most formats to simpler versions.

Reply:
Yeah, I guess with your overall rationale in the other comments, I can see why you would want to include other formats here.